src/core/expr/head_reduce_unary.cc

```diff
@@ -183,248 +183,6 @@ static Column compute_gfirstlast(Column&& arg, const Groupby&) {
 
 
 
-//------------------------------------------------------------------------------
-// product(A)
-//------------------------------------------------------------------------------
-
-template <typename T, typename U>
-bool prod_reducer(const Column& col, size_t i0, size_t i1, U* out) {
-  U prod = 1;
-  for (size_t i = i0; i < i1; ++i) {
-    T value;
-    bool isvalid = col.get_element(i, &value);
-    if (isvalid) {
-      prod *= static_cast<U>(value);
-    }
-  }
-  *out = prod;
-  return true;  // *out is not NA
-}
-
-
-
-template <typename T, typename U>
-static Column _prod(Column&& arg, const Groupby& gby) {
-  return Column(
-          new Latent_ColumnImpl(
-            new Reduced_ColumnImpl<T, U>(
-                 stype_from<U>, std::move(arg), gby, prod_reducer<T, U>
-            )));
-}
-
-static Column compute_prod(Column&& arg, const Groupby& gby) {
-  switch (arg.stype()) {
-    case SType::VOID:    return Const_ColumnImpl::make_int_column(
-                           gby.size(), 1, SType::INT64
-                         );
-    case SType::BOOL:
-    case SType::INT8:    return _prod<int8_t, int64_t>(std::move(arg), gby);
-    case SType::INT16:   return _prod<int16_t, int64_t>(std::move(arg), gby);
-    case SType::INT32:   return _prod<int32_t, int64_t>(std::move(arg), gby);
-    case SType::INT64:   return _prod<int64_t, int64_t>(std::move(arg), gby);
-    case SType::FLOAT32: return _prod<float, float>(std::move(arg), gby);
-    case SType::FLOAT64: return _prod<double, double>(std::move(arg), gby);
-    default: throw _error("prod", arg.stype());
-  }
-}
-
-
-//------------------------------------------------------------------------------
-// product(A:grouped)
-//------------------------------------------------------------------------------
-
-// T - type of input elements in the `arg` column
-// U - type of output elements
-template <typename T, typename U>
-class ProdGrouped_ColumnImpl : public Virtual_ColumnImpl {
-  private:
-    Column arg;
-    Groupby groupby;
-
-  public:
-    ProdGrouped_ColumnImpl(Column&& col, const Groupby& grpby)
-      : Virtual_ColumnImpl(grpby.size(), stype_from<U>),
-        arg(std::move(col)),
-        groupby(grpby)
-        {}
-
-    ColumnImpl* clone() const override {
-      return new ProdGrouped_ColumnImpl<T, U>(Column(arg), groupby);
-    }
-
-    size_t n_children() const noexcept override {
-      return 1;
-    }
-
-    const Column& child(size_t i) const override {
-      xassert(i == 0);  (void)i;
-      return arg;
-    }
-
-    bool get_element(size_t i, U* out) const override {
-      T value;
-      size_t i0, i1;
-      groupby.get_group(i, &i0, &i1);
-      bool isvalid = arg.get_element(i, &value);
-      *out = isvalid? static_cast<U>(i1 - i0) * static_cast<U>(value)
-                    : U(0);
-      return true; // *out is never an NA
-    }
```

Comment:
Btw, for `prod()` it seems that we had a bug in the grouped case (because it is just a copy-paste from the grouped sum implementation):
```
>>> DT
   |    C0
   | int32
-- + -----
 0 |     2
 1 |     2
 2 |     2
[3 rows x 1 column]
>>> DT[:, dt.prod(f.C0)]
   |    C0
   | int64
-- + -----
 0 |     8
[1 row x 1 column]
>>> DT[:, dt.prod(f.C0), by(f.C0)]
   |    C0     C1
   | int32  int64
-- + -----  -----
 0 |     2      6
[1 row x 2 columns]
```

Reply:
It should be `8` in the both cases, but in the last case it calculated the sum instead of the product.

Reply:
So here we actually supposed to have a shortcut: the value ^ number of elements in the group.